corehq/form_processor/models.py

```diff
@@ -551,20 +551,51 @@ def get_xml(self):
     def xml_md5(self):
         return self.get_attachment_meta('form.xml').content_md5()
 
-    def archive(self, user_id=None):
-        if self.is_archived:
-            return
+    def archive(self, user_id=None, retry_archive=False):
+        if not retry_archive:
+            # If this archive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
+            if self.is_archived:
+                return
         from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
-        FormAccessorSQL.archive_form(self, user_id=user_id)
-        xform_archived.send(sender="form_processor", xform=self)
-
-    def unarchive(self, user_id=None):
-        if not self.is_archived:
-            return
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True) as archive_stub:
+            FormAccessorSQL.archive_form(self, user_id, archive_stub)
+            archive_stub.archive_history_updated()
+            xform_archived.send(sender="form_processor", xform=self)
+
+    def unarchive(self, user_id=None, retry_archive=False):
+        if not retry_archive:
+            # If this unarchive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
+            if not self.is_archived:
+                return
+        from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False) as archive_stub:
+            FormAccessorSQL.unarchive_form(self, user_id, archive_stub)
+            archive_stub.archive_history_updated()
+            xform_unarchived.send(sender="form_processor", xform=self)
+
+    def send_archive_to_kafka(self, user_id):
```

Comment:
I don't think these functions are necessary, you can just call `xform_archived.send(sender=sender, xform=self)` in the task (assuming we agree on previous comments)